daft/table/table_io.py

```diff
@@ -102,28 +92,35 @@ def read_parquet(
     Returns:
         Table: Parsed Table from Parquet
     """
-    with _get_file(file, fs) as f:
-        pqf = papq.ParquetFile(f)
-        # If no rows required, we manually construct an empty table with the right schema
-        if read_options.num_rows == 0:
-            arrow_schema = pqf.metadata.schema.to_arrow_schema()
-            table = pa.Table.from_arrays([pa.array([], type=field.type) for field in arrow_schema], schema=arrow_schema)
-        elif read_options.num_rows is not None:
-            # Read the file by rowgroup.
-            tables = []
-            rows_read = 0
-            for i in range(pqf.metadata.num_row_groups):
-                tables.append(pqf.read_row_group(i, columns=read_options.column_names))
-                rows_read += len(tables[i])
+    if not isinstance(file, (str, pathlib.Path)):
+        # BytesIO path.
+        return Table.from_arrow(papq.read_table(file, columns=read_options.column_names))
+
+    paths, fs = _resolve_paths_and_filesystem(file, fs)
+    assert len(paths) == 1
+    path = paths[0]
+    fragment = pads.ParquetFileFormat().make_fragment(path, filesystem=fs)
```

Comment:
Qq: what motivates this change of API from `ParquetFile` to `ParquetFileFormat()`?

Reply:
Ah yes, I was going to leave comment on the diff detailing that!

pyarrow supports streaming reads (`open_input_stream`) and random access reads (`open_input_file`), where the former supports streaming decompression while the latter does not. pyarrow's fragments API (here, a `ParquetFileFragment`) supports streaming reads, while `ParquetFile` only supports random access reads. Since our previous use of `ParquetFile` was only needed because we were doing limit pushdowns (`read_options.num_rows`) by reading one row group at a time, I ported the limit pushdown to the fragment streaming read API, where we're only reading a batch of a row group of a file at a time and short-circuit as soon as we reach the row limit (which is even better than reading a row group at a time). This way, we can use the streaming read API (`open_input_stream`) for all I/O APIs.